### PR TITLE
executor: fix duplicate row check when chunk size is small

### LINKED ISSUE #7380: Wrongly throw error "Duplicate entry"
This case is taken from the [UT](https://github.com/pingcap/tidb/blob/master/executor/write_test.go#L585):
```sql
drop table if exists tmp, m;
create table tmp (id int auto_increment, code int, primary key(id, code));
create table m (id int primary key auto_increment, code int unique);
insert tmp (code) values (1);
insert tmp (code) values (1);
```

Now, let's set chunk size to `1` and run the real test query:
```sql
TiDB(localhost:4000) > set @@tidb_max_chunk_size=1;
Query OK, 0 rows affected (0.01 sec)

TiDB(localhost:4000) > insert m (code) select code from tmp on duplicate key update code = values(code);
ERROR 1062 (23000): Duplicate entry '1' for key 'code'
```

The above error **`ERROR 1062 (23000): Duplicate entry '1' for key 'code'`** should not happen, in MySQL:
```sql
MySQL(localhost:3306) > insert m (code) select code from tmp on duplicate key update code = values(code);
Query OK, 1 row affected (0.00 sec)
Records: 2  Duplicates: 0  Warnings: 0

MySQL(localhost:3306) > select * from m;
+----+------+
| id | code |
+----+------+
|  1 |    1 |
+----+------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
1. Fix #7380 , batch-get cannot get the dirty row inside the statement. It will miss the check of the duplicate keys when inserting many rows in one statement when max chunk size is small.
The old implement is used to reduce memory usage in batch-insert, but it is not the root cause. Now we can simply roll back the modification.
1. Check the row count limit for batch-insert in select execution as batch-delete does. It is the same effort as splitting the insert execution by max chunk size, which can also reduce the memory usage in batch-insert.
BTW: the batch limit check is only working for `insert into select from` now, due to it seems no one will write a 300k keys SQL using `insert into values`.

### What is changed and how it works?
Bug fix.

### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

Related changes

PTAL @coocood @tiancaiamao

### CODE REVIEW

File: executor/insert_common.go [hunk 1]
```diff
@@ -258,18 +264,31 @@ func (e *InsertValues) insertRowsFromSelect(ctx context.Context, cols []*table.C
 		}
 
 		for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
-			innerRow := innerChunkRow.GetDatumRow(fields)
+			innerRow := types.CopyRow(innerChunkRow.GetDatumRow(fields))
```
Comment (winkyao) [inline executor/insert_common.go]: Why using `types.CopyRow`? GetDatumRow is deep copied already.

**Comment by winkyao** [comment]
I guess the root cause is that in BatchGetValues, we don't get key from stmt's membuffer. It just gets MemBuffer from the tikvTxn. Before the stmtCommit is called, the kvs in stmt's membuffer will not flush to tikvTnx's Membuffer @jackysp

**Comment by winkyao** [comment]
@tiancaiamao PTAL

**Comment by winkyao** [comment]
This PR can work because of the `e.fillBackKeys` function.

File: executor/insert_common.go [hunk 2]
```diff
@@ -258,18 +264,31 @@ func (e *InsertValues) insertRowsFromSelect(ctx context.Context, cols []*table.C
...
 			row, err := e.fillRowData(cols, innerRow)
 			if err != nil {
 				return errors.Trace(err)
 			}
 			rows = append(rows, row)
+			if batchInsert && e.rowCount%uint64(batchSize) == 0 {
+				if err := exec(rows); err != nil {
+					return errors.Trace(err)
+				}
+				e.ctx.StmtCommit()
```
Comment (tiancaiamao) [inline executor/insert_common.go]: This operation will leave dirty data to binlog if error happen later.
We should never call `StmtCommit` before a statement is finished with no error.

**Comment by tiancaiamao** [comment]
LGTM

I'm ok with this fix. I think what we really need is a refactor, there are too many level for the writing path, making the code difficult to maintain. @winkyao

**Pull request merged**